chore: Cleanup of repo and dependencies [WPB-8619]

svitovyda opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-8619" title="WPB-8619" target="_blank"><img alt="Task" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10818?size=medium" />WPB-8619</a>  Cleanup of repo and dependencies
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

## Description
- add `.idea` fully to `.gitignore`;
- fix the `yarn` version in `package.json` `engines` to 4, as it's set in `packageManager` to v4 already;
- fix version of `Node` to `>=18.16` as it's the version under which unit tests have no issue with `crypto`;
- remove explicit `terser` dependency (as `webpack` has it under the hood);
- remove `exclude dexie` from minimizer `webpack` config, as `terser` has no issues with it (unlike `uglifier`);
- remove the `terser` configuration completely from `webpack`, as we don't need anything different from the default `webpack` minimization options;
- ~~add `DeadCodePlugin` to the dev `webpack` configuration;~~
- fix all dependencies versions in `package.json` to be strict (not to use `^` nor `~`), as this is our agreement;
- run `yarn up` (`yarn upgrade` in `yarn` v1) - there is a risk this and previous change erases some latest `dependabot` fixes, but it's nicer to upgrade some dependencies minor/patch versions manually than commit something in the `yarn.lock`;
- downgrade the version of `core-js` from 3.36.1 (4,323,492 downloads, month ago) to 3.26.1 (431,847 downloads, year ago) and set this version strictly in `babel` config and in `resolutions` to make sure we fully control it;
- upgrade the minimal supported versions of browsers to `99, not dead` (March-April 2022) - agreed with Sascha, exclude `opera` support (we will change it further in the other ticket, but for now, we agreed v99 can be done safely);
- ~~add instructions to the `README.md` on how to test locally the production build;~~

## Checklist

- [x] PR has been self reviewed by the author;

### Important details for the reviewers

- can be reviewed commit-by-commit
- be sure to look at comments in this PR for the output of `DeadCodePlugin` and tell if some files should be ignored as false-positives

Commit:
chore: [WPB-8619] Add .idea to .gitignore

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/wireapp/wire-webapp?pullRequest=17241) <br/>All committers have signed the CLA.

Commit:
Fix yarn version in package.json and remove explicit terser dependency

Commit:
Remove exclude dexie from terser config

Commit:
Remove completely explicit configuration of terser, as webpack does it under the hood by default;
Add DeadCodePlugin to the dev webpack configuration;

Commit:
Resolve conflicts and run `yarn up`

Commit:
Set minimal browsers support to 99+, remove Opera support;
Downgrade used `core-js` version to one year ago and make sure only this version is always used;

Commit:
Update README with instructions on how to test production build locally

svitovyda commented:
Files, identified by `webpack` `DeadCodePlugin` as not used:
```
/wire-webapp/src/script/event/CallingEvent.ts
/wire-webapp/src/script/event/EventProcessor.ts
/wire-webapp/src/script/extension/GiphyResult.ts
/wire-webapp/src/script/integration/ServiceTag.ts
/wire-webapp/src/script/util/PeerConnectionUtil.ts
/wire-webapp/src/script/util/PopupUtil.ts
/wire-webapp/src/script/view_model/CallingViewModel.mocks.ts
/wire-webapp/src/script/components/AppContainer/index.ts
/wire-webapp/src/script/components/Checkbox/Checkbox.styles.ts
/wire-webapp/src/script/components/Checkbox/Checkbox.tsx
/wire-webapp/src/script/components/Checkbox/index.ts
/wire-webapp/src/script/components/LoadingBar/index.ts
/wire-webapp/src/script/components/ServiceList/index.ts
/wire-webapp/src/script/components/calling/DeviceToggleButton.tsx
/wire-webapp/src/script/storage/record/GroupIdRecord.ts
/wire-webapp/src/script/auth/util/test/TestUtil.tsx
/wire-webapp/src/script/auth/util/test/mockStoreFactory.ts
/wire-webapp/src/script/components/MessagesList/Message/types.ts
/wire-webapp/src/script/components/Modals/LegalHoldModal/index.ts
/wire-webapp/src/script/components/panel/SingleAction/index.ts
/wire-webapp/src/script/event/preprocessor/EventStorageMiddleware/types.ts
/wire-webapp/src/script/util/test/mock/LocalStorageMock.ts
/wire-webapp/src/script/util/test/mock/ResponseMock.ts
/wire-webapp/src/script/util/test/mock/SVGProviderMock.ts
/wire-webapp/src/script/util/test/mock/WebRTCMock.ts
/wire-webapp/src/script/util/test/mock/animationMock.ts
/wire-webapp/src/script/util/test/mock/createObjectURLMock.ts
/wire-webapp/src/script/util/test/mock/cryptoMock.ts
/wire-webapp/src/script/util/test/mock/fetchMock.ts
/wire-webapp/src/script/util/test/mock/matchMediaMock.ts
/wire-webapp/src/script/util/test/mock/mediaDevicesMock.ts
/wire-webapp/src/script/util/test/mock/mutationObserverMock.ts
/wire-webapp/src/script/util/test/mock/navigatorPermissionsMock.ts
/wire-webapp/src/script/util/test/mock/resizeObserver.mock.ts
/wire-webapp/src/script/page/RightSidebar/ConversationDetails/components/GroupDetails/index.ts
```

atomrc commented on .gitignore:
```diff
@@ -15,6 +13,7 @@ CHANGELOG.md
 node_modules
 npm-debug.log
 yarn-error.log
+*.dump
```
please add those kind of files to your personal global `.gitignore`. They don't belong in the repo directly

atomrc commented on .gitignore:
```diff
@@ -23,4 +22,5 @@ yarn-error.log
-!.yarn/versions
\ No newline at end of file
+!.yarn/versions
+/wire-webapp.iml
```
please add those kind of files to your personal global `.gitignore`. They don't belong in the repo directly

atomrc commented on webpack.config.dev.js:
```diff
@@ -34,7 +36,14 @@ module.exports = {
+    new DeadCodePlugin({
+      patterns: ['src/**/*.(js|jsx|css|ts|tsx)'],
+      exclude: ['**/*.(stories|spec|test).(js|jsx|ts|tsx)', 'src/types/*.*', '**/*.d.ts', '**/__mocks__'],
+    }),
```
I don't think we need to add it at this point. This might slow down dev workflow a bit and we almost never look at the webpack report anyway. 
It's okay if files are not consumed (and actually there are a lot of false positive in the report it generated, like mock for testing)

svitovyda replied:
Should I just create a Jira for removing the unused code using this report (excluding mocks)?

atomrc commented on README.md:
```diff
@@ -52,6 +52,18 @@ If you would like your browser to trust the certificate from "local.wire.com" or
+1. `cp ../../.env .`
+1. Edit `.env`, add `BRAND_NAME=Wire`
+1. `node index.js`
+
```
This is a very hacky way of running the server. We should not advertize that publicly in our README :)

Commit:
Move some ignored patterns from project .gitignore to the global one;
Remove DeadCodePlugin;
Remove internal instructions from public README;

Commit:
Bump back the version of `core-js` to 3.36.1

tlebon commented on package.json:
```diff
@@ -150,8 +149,8 @@
-    "yarn": ">= 1.0.0",
-    "node": ">=16"
+    "yarn": ">= 4.1.1",
+    "node": ">=18.16.0"
```
still not sure what happens with this runs on our *unfortunately* still node 16 EB instance

atomrc replied:
I don't think it's a problem, since the aws never runs anything else than expressjs. 
All the rest is just for CI and local dev, we will be fine I believe

tlebon commented on webpack.config.js:
```diff
@@ -27,12 +26,7 @@ module.exports = {
-        exclude: /dexie/g,
-      }),
-    ],
+    minimize: true,
```
this doesnt cause any issues?

svitovyda replied:
no, I tested it locally with production version, even with no sourceMap to be sure. tried sending message, deleting, saving backup, recovering from backup.

atomrc reviewed (approved):
Looks solid 💪😎 

Nice first PR 🥳 🎉
